### PR TITLE
support multi-schema-change for drop index

### LINKED ISSUE #21: Support multi-schema change for `drop index`

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #21

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -701,6 +686,44 @@ func onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+		}
+	}
+	// Sort the offset in descending order.
+	sort.Slice(hiddenColOffs, func(i, j int) bool {
+		return hiddenColOffs[i] > hiddenColOffs[j]
+	})
+	// Move all the dependent hidden columns to the end.
+	endOffset := len(tblInfo.Columns) - 1
+	for _, offset := range hiddenColOffs {
+		tblInfo.MoveColumnInfo(offset, endOffset)
```
Comment (Defined2014) [inline ddl/index.go:704]: Seems `slice` is not a good structure for column info. Linked list is better. But this will affect many files, 😭

Comment (Defined2014) [inline ddl/index.go:705]:
```suggestion
		tblInfo.MoveColumnInfo(offset, endOffset)
		endOffset--
```

File: ddl/index.go [hunk 2]
```diff
@@ -701,6 +686,44 @@ func onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+func removeDependentHiddenColumns(tblInfo *model.TableInfo, idxInfo *model.IndexInfo) {
+	hiddenColOffs := make([]int, 0)
+	for _, indexColumn := range idxInfo.Columns {
+		col := tblInfo.Columns[indexColumn.Offset]
+		if col.Hidden {
+			hiddenColOffs = append(hiddenColOffs, col.Offset)
+		}
+	}
+	// Sort the offset in descending order.
+	sort.Slice(hiddenColOffs, func(i, j int) bool {
```
Comment (Defined2014) [inline ddl/index.go:699]: Maybe we could reimplement this without sort. Like 
```
for _, offset := range hiddenColOffs {
    if offset >= endOffset {
        continue
    }
    for {
        if needDelete(tblInfo.Columns[offset]) {
            tblInfo.MoveColumnInfo(offset, endOffset)
            endOffset--
        } else {
            break
        }
    }
}
```

Comment (Defined2014) [reply ddl/index.go:699]: By the way, the difference is not too much.

File: ddl/multi_schema_change_test.go [hunk 3]
```diff
@@ -128,3 +128,17 @@ func TestMultiSchemaChangeOperateSameColumn(t *testing.T) {
 	_, err = tk.Exec("alter table t drop column a, drop column a")
 	require.Equal(t, err.Error(), "[ddl:8200]Unsupported operate same column 'a'")
 }
+
+func TestMultiSchemaChangeDropIndexes(t *testing.T) {
```
Comment (Defined2014) [inline ddl/multi_schema_change_test.go]: Need rebase, this func name already exists in `multi-schema-change` branch.

File: ddl/index.go (hunk 1 above)
Comment (tangenta) [reply ddl/index.go:705]: I think the order does not matter, because the columns will be removed anyway.

Comment (Defined2014) [reply ddl/index.go:705]: It will reduce the number of exchanges

File: ddl/multi_schema_change_test.go [hunk 4]
```diff
@@ -128,3 +128,17 @@ func TestMultiSchemaChangeOperateSameColumn(t *testing.T) {
...
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test;")
+	tk.MustExec("set @@global.tidb_enable_change_multi_schema = 1;")
+
+	tk.MustExec("create table t (id int, c1 int, c2 int, primary key(id), key i1(c1), key i2(c2));")
+	tk.MustExec("insert into t values (1, 2, 3);")
+	tk.MustExec("alter table t drop index i1, drop index i2;")
+	tk.MustExec("select * from t use index(i1);")
```
Comment (Defined2014) [inline ddl/multi_schema_change_test.go]:
```suggestion
	tk.MustGetErrCode("select * from t use index(i1);", errno.ErrKeyDoesNotExist)
```

File: ddl/multi_schema_change_test.go [hunk 5]
```diff
@@ -128,3 +128,17 @@ func TestMultiSchemaChangeOperateSameColumn(t *testing.T) {
...
+	tk.MustExec("select * from t use index(i2);")
```
Comment (Defined2014) [inline ddl/multi_schema_change_test.go]:
```suggestion
	tk.MustGetErrCode("select * from t use index(i2);", errno.ErrKeyDoesNotExist)
```

File: ddl/column.go [hunk 6]
```diff
@@ -217,6 +217,9 @@ func onAddColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error)
 }
 
 func locateOffsetForColumn(pos *ast.ColumnPosition, tblInfo *model.TableInfo) (offset int, err error) {
+	if pos == nil {
+		return -1, nil
+	}
```
Comment (tangenta) [inline ddl/column.go:222]: `locateOffsetForColumn` is also used by `AddColumn` so the pos can be null.

File: ddl/column.go [hunk 7]
```diff
@@ -540,10 +543,7 @@ func onDropColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	}
 	if job.MultiSchemaInfo != nil && job.MultiSchemaInfo.Revertible {
 		job.MarkNonRevertible()
-		ver, err = updateVersionAndTableInfoWithCheck(t, job, tblInfo, false)
-		if err != nil {
-			return ver, errors.Trace(err)
-		}
+		return updateVersionAndTableInfoWithCheck(t, job, tblInfo, false)
```
Comment (tangenta) [inline ddl/column.go:546]: Added tests `TestMultiSchemaChangeAddColumns`:

```
create table t (a int, b int, c int);
alter table t add column (d int default 4, e int default 5);
```

File: ddl/column.go [hunk 8]
```diff
@@ -362,7 +365,7 @@ func onAddColumns(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error
 	case model.StateWriteReorganization:
 		// reorganization -> public
 		// Adjust table column offsets.
-		for i, newCol := range tblInfo.Columns[:len(tblInfo.Columns)-len(positions)] {
+		for i, newCol := range tblInfo.Columns[len(tblInfo.Columns)-len(positions):] {
```
Comment (tangenta) [inline ddl/column.go:368]: Added test in `TestMultiSchemaChangeAddColumns`:

```
create table t (a int, b int, c int);
alter table t add column (d int default 4, e int default 5);
```

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**